# Analysis notebook

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Import your helper module
import sys
sys.path.insert(0, '..')
from src.helpers import *

How long did the avg project take
was there more residential development in __ or ___
Are applications being processed faster now than they were?


In [2]:
# Error message when loading data as usually, converted to str for now
df = pd.read_csv('../data/building-permits.csv')

C:\Users\aylie\AppData\Local\Temp\ipykernel_35984\4097426922.py:2: DtypeWarning: Columns (23,24,25,26) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../data/building-permits.csv')


## Initial Exploration

### Data Shape

In [3]:
df.shape

(157928, 35)

*35 Columns*<br>
*157928 Rows*

### Data Information

This dataset has 35 columns. As such, the outputs of functions that return information for each column are truncated due to length. Instead of running them directly in the notebook they are nested in helper functions which return a cleaner output containing the same information:

**Data Types**

In [4]:
# Point at src folder for helper functions
import sys
sys.path.append('../src')
from helpers import split_columns

# Get columns and data type in table format (2 cols)
split_columns(df)


,Column Names,Data Types,Column Names (cont.),Data Types (cont.)
0,issue_date,object,dwelling_units_lost,int64
1,permit_number,object,location,object
2,parent_permit_number,object,x_coordinate_nad83,float64
3,permit_group,object,y_coordinate_nad83,float64
4,permit_type,object,includes_secondary_suite,object
5,sub_type,object,adding_secondary_suite,object
6,work_type,object,removing_secondary_suite,object
7,street_number,float64,pool_type,object
8,street_name,object,type_of_structure,object
9,street_type,object,ward,object


In [13]:
# Use .dtypes and .value_counts() to get count of data types
df.dtypes.value_counts()

object     29
float64     4
int64       2
Name: count, dtype: int64

*object:* 29 <br>
*float:* 4 <br>
*integer:* 2

**NaN Counts**

In [ ]:
# Point at src folder for helper functions
sys.path.append('../src')
from helpers import nan_count

# Call nan_count() function to get count of NaN values in each column
nan_count(df)

,Column Names,NaN Count,Column Names (cont.),NaN Count (cont.)
0,issue_date,0,dwelling_units_lost,0.0
1,permit_number,0,location,355.0
2,parent_permit_number,145795,x_coordinate_nad83,355.0
3,permit_group,0,y_coordinate_nad83,355.0
4,permit_type,0,includes_secondary_suite,111342.0
5,sub_type,0,adding_secondary_suite,144047.0
6,work_type,2,removing_secondary_suite,144047.0
7,street_number,290,pool_type,156825.0
8,street_name,272,type_of_structure,156967.0
9,street_type,1126,ward,88.0


*Observations about NaN counts:*
1. counts for 'location', 'x_coordinate_nad83', 'y_coordinate_nad83', and 'point' are the same.
2. counts for 'adding_secondary_suite' and 'removing_secondary_suite' are the same.
3. counts for 'unit_type' and 'unit_number' are the same.
4. counts for 'neighbourhood_number' and 'neighbourhood_name' are the same.
5. counts  for 'work_type' and 'address' are the same.

There are three categories of NaN values present in this data: 
1. Expected (where the NaN can be interpreted as 'N/A')
2. Innocuous (NaN value is not a cause for concern)
3. Potentially harmful (NaNs that need to be handled and may signal data quality issues - ex: NaN in what would have been a primary key column)

The table below shows where the columns in this dataset fall within thesed three categories:

| Expected (N/A) | Innocuous | Potentially Harmful |
|---|---|---|
| parent_permit_number | ward | work_type |
| street_direction | | street_number |
| unit_type | | street_name |
| unit_number | | street_type |
| applicant_business_name | | neighbourhood_name |
| pool_type | | neighbourhood_number |
| | | location |
| | | x_coordinate_nad83 |
| | | y_coordinate_nad83 |
| | | type_of_structure |
| | | final_date |
| | | point |
| | | address |

Many of these can be filled in at a later state. For example, 'street_number', 'street_name', 'street_type', and 'street_direction' are all atomic versions of the 'address' column. While the atomic versions have many missing values, the address column only has 2:

In [19]:
# Display NaN count for columns relating to address
nan_count(df[['street_number', 'street_name', 'street_type', 'street_direction', 'address']])

,Column Names,NaN Count,Column Names (cont.),NaN Count (cont.)
0,street_number,290,street_direction,149811.0
1,street_name,272,address,2.0
2,street_type,1126,,


In [ ]:
# Display a row of columns pertaining to address to show relation
df[['street_number', 'street_name', 'street_type', 'street_direction', 'address']].iloc[7]

street_number             94.0
street_name               Hill
street_type                 ST
street_direction           NaN
address             94 Hill ST
Name: 7, dtype: object